contracts/interchain-token-service/src/state.rs

```diff
@@ -107,6 +180,104 @@ pub fn load_all_its_contracts(
         .collect::<Result<HashMap<_, _>, _>>()?)
 }
 
+pub fn save_token_info(
+    storage: &mut dyn Storage,
+    directional_chain: DirectionalChain,
+    token_id: TokenId,
+    token_deployment_type: TokenDeploymentType,
+) -> Result<(), Error> {
+    let token_type = TokenType::from(&directional_chain);
+    let chain: ChainNameRaw = directional_chain.into();
+
+    match TOKEN_INFO.may_load(storage, &(chain.clone(), token_id.clone()))? {
+        Some(_) if token_type == TokenType::Origin => (),
+        Some(_) => return Err(Error::TokenAlreadyDeployed { token_id, chain }),
+        None => {
+            let balance = (token_type.clone(), token_deployment_type).into();
+
+            TOKEN_INFO.save(
+                storage,
+                &(chain, token_id),
+                &TokenInfo {
+                    balance,
+                    token_type,
+                },
+            )?;
+        }
+    }
+
+    Ok(())
+}
+
+pub fn update_token_info(
+    storage: &mut dyn Storage,
+    directional_chain: DirectionalChain,
+    token_id: TokenId,
+    amount: nonempty::Uint256,
+) -> Result<(), Error> {
+    let chain: ChainNameRaw = directional_chain.clone().into();
+    let Some(mut token_config) =
+        TOKEN_INFO.may_load(storage, &(chain.clone(), token_id.clone()))?
+    else {
+        return Err(Error::TokenNotDeployed { token_id, chain });
+    };
+
+    token_config.balance = match directional_chain {
+        DirectionalChain::Source(_) => token_config.balance.checked_sub(amount),
+        DirectionalChain::Destination(_) => token_config.balance.checked_add(amount),
+    }
+    .map_err(|_| Error::TokenBalanceInvariantViolated {
+        token_id: token_id.clone(),
+        chain: chain.clone(),
+    })?;
+
+    TOKEN_INFO.save(storage, &(chain, token_id), &token_config)?;
```

Comment:
I think you could use `update` for this: https://docs.rs/cw-storage-plus/latest/cw_storage_plus/struct.Map.html#method.update